### PR TITLE
Basefinder - Optimization and Bugfixing

### PR DESCRIPTION
**Pull request by ghost**
Bug:
"Block.getBlockFromName("cobweb")"
will return null since the correct name of the block is "web"

### CODE REVIEW

File: src/tk/wurst_client/mods/BaseFinderMod.java [hunk 1]
```diff
@@ -60,94 +69,51 @@ public void onUpdate()
...
 			for(int y = range; y >= -range; y--)
-			{
 				for(int x = range; x >= -range; x--)
-				{
 					for(int z = range; z >= -range; z--)
 					{
-						int posX = (int)(mc.thePlayer.posX + x);
-						int posY = (int)(mc.thePlayer.posY + y);
-						int posZ = (int)(mc.thePlayer.posZ + z);
+						int posX = playerX + x;
```
Comment (CisBetter) [inline src/tk/wurst_client/mods/BaseFinderMod.java:81]: I'd move the declaration of posX, posY, and posZ out of the `for` loop:

``` Java
int playerZ = (int) Math.floor(mc.thePlayer.posZ);
int posX;
```

...

``` Java
for(int z = range; z >= -range; z--)
{
    posX = playerX + x;
```

File: src/tk/wurst_client/mods/BaseFinderMod.java [hunk 2]
```diff
@@ -60,94 +69,51 @@ public void onUpdate()
...
+					shouldInform = false;
+				}
+			} else
 				shouldInform = true;
+			
 			updateLastMS();
 		}
 	}
 	
+	private boolean foundTooManyBlocks() {
```
Comment (CisBetter) [inline src/tk/wurst_client/mods/BaseFinderMod.java:111]: Why would you need a function for that? Every function call takes time for the JVM to execute. I'd suggest replacing `foundTooManyBlocks()` with `(matchingBlocks.size() >= maxBlocks)`.
If I've missed something let me know :smile:

**Review by CisBetter — CHANGES_REQUESTED** [review]

File: src/tk/wurst_client/mods/BaseFinderMod.java (hunk 2 above)
Comment (ghost) [reply src/tk/wurst_client/mods/BaseFinderMod.java:111]: Most JIT comilers start optimizing at ~10 000 calls.
This will be reached before even one tick passed.

So I would expect that it will be inlined at runtime.
You can run some benchmarks...

File: src/tk/wurst_client/mods/BaseFinderMod.java (hunk 1 above)
Comment (ghost) [reply src/tk/wurst_client/mods/BaseFinderMod.java:81]: I don't like uninitialized values, so I would like to keep them in the loop.
Moving them out of the loop would add 3 (pointless (my opinion)) additional lines.

File: src/tk/wurst_client/mods/BaseFinderMod.java (hunk 2 above)
Comment (ghost) [reply src/tk/wurst_client/mods/BaseFinderMod.java:111]: I measured the time the loops took to iterate and was not able to measure any conspicuous speed differences

Comment (CisBetter) [reply src/tk/wurst_client/mods/BaseFinderMod.java:111]: Perhaps it's just because I mainly program in `C` but I still wonder why one should rely on something **if they don't have to.** This function doesn't make the code faster nor easier to read. But once again, I'm not so comfortable with the JVM, nor am I comfortable with Java compilers. If I'm totally wrong please let me know.

**Review by ghost — COMMENTED** [review]
@CisBetter  The original code was using the code 5 times so I put it into a seperate function.
It's only used twice now, so we can inline it, if you really want to.

File: src/tk/wurst_client/mods/BaseFinderMod.java (hunk 1 above)
Comment (CisBetter) [reply src/tk/wurst_client/mods/BaseFinderMod.java:81]: As the variables aren't used outside of the loop there is technically no way they could be used without having a value assigned to them. There is a difference between a variable that is created and destroyed 1024 times (total: **2048** times) and a variable that's been created and destroyed 1 time and that has been assigned 1024 times (total: **1026** times). Even though Java is so ridiculously slow and you won't notice a difference in execution time in I'd call it `good coding style`.

But if you want to you can leave it like that as the compiler will move it out of the loop any ways (at least it should). Keeping it like that is much easier to maintain.

**Review by CisBetter — APPROVED** [review]

File: src/tk/wurst_client/mods/BaseFinderMod.java (hunk 1 above)
Comment (ghost) [reply src/tk/wurst_client/mods/BaseFinderMod.java:81]: "There is a difference between a variable that is created and destroyed 1024 times"
(btw: the loop iterates 100^3 times)
That's absolutly true, but only for fields (members of classes).

Local variables (variables inside of functions) aren't created and destroyed everytime you declare the in java source.
tl;dr: Both versions will Compile to the same bytecode.

longer version:
local variables are represented in jvm bytecode like an array.
This array is created as soon as the method is entered and will get destroyed if the method is exited.
The compiler will assign an index for each local variable, so the size of the array is the local variable count.
jvm bytecode does not have local variable declarations.

So if we take the following code:

```
public static void foo() {
    for(int i = 0; i < 10; i ++) {
        int x = 20;
        int y = 30;
        int z = 40;
        System.out.println(x + y + z);
    }
}
```

You can say (simplified) that it will compile to something like that:

```
public static void foo() {
    Object[] locals = new Object[4];
    for(locals[0] = 0; locals[0] < 10; locals[0]++) {
        locals[1] = 20;
        locals[2] = 30;
        locals[3] = 40;
        System.out.println(locals[1] + locals[2] + locals[3]);
    }
}
```

Now you should be able to understand that there's not difference.

Comment (ghost) [reply src/tk/wurst_client/mods/BaseFinderMod.java:81]: btw: **JAVA IS NOT SLOW!**
(It's really annoying to hear that myth over and over again)

It can almost be as fast as C/C++ and sometime even faster!
Modern jvm's have very good just in time compilers.
They will do optimization based on runtime profiling.

e.g. Escaped analysis allows to detect pure data structures classes.
The jvm won't create instances of those and pass the values directly to the corresponding place.
Or if the condition of an if-statement has never evaluated to true during many executions, the jvm can make some further optimizations.

That's not possible with compilers like c and c++ since they produce one static output file.

Comment (MatMercer) [reply src/tk/wurst_client/mods/BaseFinderMod.java:81]: @CisBetter
In C/C++ there is the problem of allocate and deallocate memory all the time, I programmed in C a lot this year and I know why you thought that it would be a performance issue in this for loop.

But (greatly) in Java is different! Like @yellowlight2 said, it doesn't deallocate in a local context. 

Another thing is that each time you do "int x;" x is assign to 0, like implicitly "int x = 0;" so neither undeclared variables would be an issue, it is just a matter of personal coding hah.
